### ISSUE TITLE
Issue with multiplication of 'large' numbers

### ISSUE THREAD

**Issue by toeklk** [gsdlab/chocosolver#10]
Consider the following code
```
abstract hisclafer
  atth -> int

abstract yourclafer
  tmpatt -> hisclafer
  att -> int = tmpatt.atth

abstract myclafer
  att1 -> yourclafer 1..3
  att2 -> int = sum att1.att
  att3 -> int = (#att1) * att2

hisclaferinstanceOne : hisclafer
  [atth = 5]

hisclaferinstanceTwo : hisclafer
  [atth = 5]

yourclaferinstanceOne : yourclafer
  [tmpatt = hisclaferinstanceOne]

yourclaferinstanceTwo : yourclafer
  [tmpatt = hisclaferinstanceTwo]

myclaferinstance : myclafer
  [att1 = yourclaferinstanceOne, yourclaferinstanceTwo]
```
which runs perfectly fine on clafer 0.3.9

```
 java -jar ~/git/gsdlab/ClaferChocoIG/target/claferchocoig-0.3.9-jar-with-dependencies.jar -n 1 --maxint 64 --prettify --file multiplication.cfr
Compiling the Clafer model...
Running Model...
Generating instances...
=== Instance 1 Begin ===

c0_hisclaferinstanceOne#0
    c0_atth#0 = 5
c0_hisclaferinstanceTwo#0
    c0_atth#1 = 5
c0_yourclaferinstanceOne#0
    c0_tmpatt#0 = c0_hisclaferinstanceOne#0
    c0_att#0 = 5
c0_yourclaferinstanceTwo#0
    c0_tmpatt#1 = c0_hisclaferinstanceTwo#0
    c0_att#1 = 5
c0_myclaferinstance#0
    c0_att1#0 = c0_yourclaferinstanceOne#0
    c0_att1#1 = c0_yourclaferinstanceTwo#0
    c0_att2#0 = 10
    c0_att3#0 = 20
--- Instance 1 End ---
```

Running the same code with clafer-0.3.10 doesn't produce any outputs at all
```
 java -jar ~/git/gsdlab/ClaferChocoIG/target/claferchocoig-0.3.10-jar-with-dependencies.jar -n 1 --maxint 64 --prettify --file multiplication.cfr
Compiling the Clafer model...
Running Model...

Generating instances...
Generated 0 instance(s)
```
It seems like, as soon 
``` att3 -> int = (#att1) * att2 > 16```

claferChocoIG fails to produce any outputs

**Comment by toeklk** [comment]
Initial analysis seems to indicate that things go wrong in the creation of the javascript file.

Indeed, both for clafer 0.3.9 and 0.3.10 the javascript file looks like
```
intRange(-8, 7);
```

no matter what is specified on the command line as --maxint.

Yet, in the case of clafer 0.3.10 this cause havoc, due to the following code in ScopeBuilder.java
```
public Scope toScope() {
        return new Scope(scope, defaultScope, intLow, intHigh,
                mulLow == null ? Math.min(-16, intLow) : mulLow,
                mulHigh == null ? Math.max(16, intHigh) : mulHigh,
                stringLength, charLow, charHigh);
``

Later more...

**Comment by toeklk** [comment]
Ok, I think I finally understood what goes wrong.  When passing a clafer file to claferchocoig-0.3.10-jar-with-dependencies.jar , the clafer compiler produces a javascript file and at that time does not take into account command line options such as maxint.  So no matter what is specified, the javascript code contains 
```
intRange(-8, 7);
```
Hower, during the reading of the model file, the above toScope method is called for the first time
```
choco-test [Java Application]	
	Main at localhost:56129	
		Thread [main] (Suspended (breakpoint at line 233 in ScopeBuilder))	
			ScopeBuilder.toScope() line: 233	
			JavascriptContext.getScope(Scriptable) line: 79	
			Javascript.readModel(String, Reader, Scriptable) line: 91	
			Javascript.readModel(String, File, Scriptable) line: 76	
			Javascript.readModel(File, Scriptable) line: 53	
			Javascript.readModel(File) line: 45	
			Normal.runNormal(File, OptionSet, PrintStream) line: 27	
			Main.main(String[]) line: 120	
	/Library/Java/JavaVirtualMachines/jdk1.8.0_40.jdk/Contents/Home/bin/java (05 May 2015 09:20:30)	
```
In this case (as intended I guess), mulLow and mulHigh are uninitialized and hence mulLow and mulHigh are set to [-16,16] (see above code in comment 1).

However, next the code reaches an explicit call to the toScope function as part of the runNormal function
```
Normal.runNormal(File, OptionSet, PrintStream) line: 53	
Main.main(String[]) line: 120	
```
If mulLow and mulHigh would still be null, then everything would work as expected.  However, they were already set as a part of the first call and hence the toScope method() is called this time with the following parameters (--maxint 255 was used in this case)
```
this	ScopeBuilder  (id=94)	
	charHigh	~	
	charLow	
	defaultScope	1	
	intHigh	255	
	intLow	-256	
	mulHigh	Integer  (id=96)	
		value	16	
	mulLow	Integer  (id=95)	
		value	-16	
	scope	HashMap<K,V>  (id=97)	
	stringLength	16	
```
So, whatever is passed at the command line, mulLow and mulHigh are always set to -16,16 (making claferChocoIG almost useless for multiplications)

I see two ways to solve this, not sure which one is to be preferred
- Either the clafer compiler should produce a javascript file with correct scope (in this case: not a chocosolver bug :-)
- Either the toScope method() should be made more robust in case it is called multiple times

Tmp workaround:
- Manually adapt the default scopes in the js file

**Comment by JLiangWaterloo** [comment]
Thanks for this very detailed bug report.

Let me provide some background, previously, the mulLow and mulHigh were never set and the domain of a multiplication is simply computed based on the domains of the multiplication's inputs. However, this caused problems when the multiplication has many inputs (for example, one of our models was multiplying 16+ integer variables), and domain of the multiplication would be very large and often times even overflow, i.e. >2^31. To fix this problem, I implemented mulLow and mulHigh so that we can configure how large multiplications can be during solving.

However, as you noted, my implementation was incorrect. I pushed a fix to the develop branch which basically sets mulLow=intLow and mulHigh=intHigh unless mulLow/mulHigh are explictly set.

I tried your example in the original post and I was able to get a solution post-fix.
